src/smt/theory_str_noodler/decision_procedure.cpp

```diff
@@ -592,96 +592,136 @@ namespace smt::noodler {
             return LenNode(LenFormulaType::TRUE);
         }
 
-        // start with length formula from preprocessing
-        std::vector<LenNode> conjuncts = {preprocessing_len_formula};
+        // start with formula for disequations
+        std::vector<LenNode> conjuncts = disequations_len_formula_conjuncts;
+        // add length formula from preprocessing
+        conjuncts.push_back(preprocessing_len_formula);
 
-        // decision procedure was run, we create length constraints from the solution, we only need to look at length sensitive vars
+        // create length constraints from the solution, we only need to look at length sensitive vars
         for (const BasicTerm &len_var : solution.length_sensitive_vars) {
             conjuncts.push_back(solution.get_lengths(len_var));
         }
 
-        // add the formula for solving disequations
-        conjuncts.push_back(diseqs_formula());
+        // the following functions (getting formula for transformations) assume that we have flattened substitution map
+        solution.flatten_substition_map();
+
+        // add formula for transformations
+        conjuncts.push_back(transformation_formula());
 
         return LenNode(LenFormulaType::AND, conjuncts);
     }
 
-    LenNode DecisionProcedure::diseqs_formula() {
-        // This function creates arithmetic formula using solution for each ((a1, a2), (|x1| == |x2|, |L| != |R|) in
-        // dis_len where L = x1a1y1, R= x2a2y2; see replace_disequality() for more info. We want to create formula
-        // (|L| != |R| AND (|x1| == |x2| OR a1 != a2)) but
-        //      - if a1 or a2 is only empty word, then only |L| != |R| can hold
-        //      - we check a1!char != a2!char instead of a1 != a2 and then we add at the end that a1!char/a2!char
-        //        should be equal to one of the symbols of a1/a2
-        //      - we also work with maximally substituted a1/a2, so that if we have multiple disequations that
-        //        somehow depend on each other, the dependency stays
-
-        // we flatten the subtitution map, so that all the vars a1/a2 whose disequality we will check
-        // are substituted by maximal application of substitution_map
-        solution.flatten_substition_map();
-
-        // here we save all a1/a2 whose disequation is checked, so that we can create the formula that 
-        // each of these vars should be one of the symbols from its automaton
-        std::set<BasicTerm> a_vars;
-        // get the "!char" version of a variable
-        auto get_char_var = [](const BasicTerm& var) -> BasicTerm { return BasicTerm(BasicTermType::Variable, var.get_name().encode() + "!char"); };
-
-        // conjuncts of the final length formula
-        std::vector<LenNode> conjuncts;
+    LenNode DecisionProcedure::transformation_formula() {
+        STRACE("str-transform",
```

Comment:
I think it is better to have only the main things in "str", and everything else should be divided. You can then enable only those traces you actually need, instead of all of them.